### ISSUE TITLE
Ground-truth camera pose is not used during training Neural Mesh Models

### ISSUE THREAD

**Issue by kvuong2711** [Angtian/NeMo#4]
Hello,

In the paper, while training the Neural Mesh Models,  _"the correspondence between the feature vector f_i in the feature map F and the vector θ_r on the neural mesh model is given by the 2D projection of the mesh with camera parameters m"_. I assume that this means during training, the neural mesh model is rendered using the ground-truth camera pose m, then compared with the features of the RGB image. 

However, I don't see this being the case in the code. Can you give some comments about the actual implementation?

Thanks in advance.

**Comment by Angtian** [comment]
Hello, 
Thanks for your interest in our work. In our implementation, the projection and visibility is conducted in the data preprocess using the groundtruth camera: https://github.com/Angtian/NeMo/blob/8875eb0e8b8e2c2a8cba68f0865c78c2e42e3242/code/dataset/generate_3Dpascal3D.py#L69
Specifically, the project is computed here:
https://github.com/Angtian/NeMo/blob/8875eb0e8b8e2c2a8cba68f0865c78c2e42e3242/code/lib/MeshMemoryMap.py#L49
The visibility is computed (where we implement the rasterization process by ourself, note it is also feasible to use pytorch3d to conduct this process):
https://github.com/Angtian/NeMo/blob/8875eb0e8b8e2c2a8cba68f0865c78c2e42e3242/code/lib/MeshMemoryMap.py#L58

**Comment by kvuong2711** [comment]
Hi,

Thank you very much for your quick response. Appreciate it! I also have another question: 

In `VisualizeUnlabeledDataset.py`, PyTorch3D's camera is `cameras = OpenGLPerspectiveCameras(device=device, fov=12.0)`. I guess `fov=12.0` is arbitrary and doesn't matter as long as this is also the same during the inference process? (Maybe not necessarily arbitrary, with width = 672, FOV=12.0 approximates to focal length ~=3000).

However, if I want to render the 3D mesh onto the image using the ground-truth camera pose for an image in PASCAL3D+, I know that the default focal length of PASCAL3D+ is 3000. However, changing to `cameras = PerspectiveCameras(focal_length=3000)` doesn't work (the rendered car is too big). Do you know if there's any setting that I missed?

In other words, is there a version of [this](https://github.com/Angtian/NeMo/blob/8875eb0e8b8e2c2a8cba68f0865c78c2e42e3242/code/lib/ProcessCameraParameters.py#L87) but in PyTorch3d?

EDIT: Actually I figured it out. Using `PerspectiveCameras()` with `in_ndc=False` works well.

Thanks!

**Comment by Angtian** [comment]
Hello,
Yes, you are correct. When we implement this, the PyTorch3D is in an early version and lacks of support for PerspectiveCameras. So we manually adapt this camera and find out fov=12 is equivalent to image_size=(256, 672) with focal length =3000.
